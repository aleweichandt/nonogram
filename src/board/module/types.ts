export type Options<T> = T[];
export type Board<T> = T[][];

export type State<T> = {
  board: Board<T>;
  currentOption: T;
  options: Options<T>;
  progress: Board<T>;
};

export type StateWithBoard<T> = {
  board: State<T>;
};
