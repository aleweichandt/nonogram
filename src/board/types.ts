export type GridProps = {
  row?: number;
  column?: number;
};

export type SelectorProps<T> = {
  value: T;
};
