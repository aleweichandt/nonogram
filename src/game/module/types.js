// @flow
import type { BoardType, OptionType, OptionsType } from '../types';

export * from '../types';

export type StateType = {
  board: BoardType,
  currentOption: OptionType,
  options: OptionsType,
  progress: BoardType,
}

export type StateWithGameType = {
  game: StateType,
}
