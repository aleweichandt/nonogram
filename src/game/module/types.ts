import {OPTION_BLACK, OPTION_BLUE, OPTION_BLOCKED, OPTION_VOID} from './const';

export type OptionType =
  | typeof OPTION_BLACK
  | typeof OPTION_BLUE
  | typeof OPTION_BLOCKED
  | typeof OPTION_VOID;

export type OptionsType = OptionType[];

export type InfoType = {
  option: OptionType;
  count: number;
  complete: boolean;
};
export type LineInfoType = InfoType[];

export type LineType = OptionType[];
export type BoardType = OptionType[][];

export type StateType = {
  board: BoardType;
  currentOption: OptionType;
  options: OptionsType;
  progress: BoardType;
};

export type StateWithGameType = {
  game: StateType;
};
