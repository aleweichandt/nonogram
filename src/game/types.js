// @flow
import { OPTION_BLACK, OPTION_BLOCKED, OPTION_VOID } from './const';

export type OptionType = |
  typeof OPTION_BLACK |
  typeof OPTION_BLOCKED |
  typeof OPTION_VOID;

export type OptionsType = OptionType[];
export type BoardType = OptionType[][];
