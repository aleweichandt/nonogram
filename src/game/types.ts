import {OptionType} from './module';

export type GridProps = {
  row?: number;
  column?: number;
};

export type SelectorProps = {
  value: OptionType;
};
