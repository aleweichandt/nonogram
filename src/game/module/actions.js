import { DEFAULT_OPTIONS, OPTION_VOID, OPTION_BLACK } from '../const';

export const INIT_GAME = 'game/INIT_GAME';
export const SET_OPTION = 'game/SET_OPTION';
export const SET_TILE = 'game/SET_TILE';
export const CHECK_STATE = 'game/CHECK_STATE';
export const END_GAME = 'game/END_GAME';

/**
 - init game
 - set_tile -> check_state -> set_state [-> end_game]
 */

export const initGame = (
  board = [[]],
  options = DEFAULT_OPTIONS,
  currentOption = OPTION_BLACK,
) => ({
  type: INIT_GAME,
  payload: {
    board,
    currentOption,
    options,
  },
});

export const setOption = (option = OPTION_VOID) => ({
  type: SET_OPTION,
  payload: {
    option,
  },
});

export const setTile = (row = 0, col = 0) => ({
  type: SET_TILE,
  payload: {
    row,
    col,
  },
});

export const checkState = () => ({
  type: CHECK_STATE,
});

export const endGame = () => ({
  type: END_GAME,
});
