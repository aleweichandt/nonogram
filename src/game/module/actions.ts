import {PayloadAction, Action} from '../../redux-helpers';
import {Board, Options} from './types';

export const INIT_GAME = 'game/INIT_GAME';
export const SET_OPTION = 'game/SET_OPTION';
export const SET_TILE = 'game/SET_TILE';
export const CHECK_STATE = 'game/CHECK_STATE';
export const END_GAME = 'game/END_GAME';

export type InitGame<T> = PayloadAction<
  typeof INIT_GAME,
  {
    board: Board<T>;
    voidOption: T | void;
    options: Options<T>;
  }
>;
export type SetOption<T> = PayloadAction<
  typeof SET_OPTION,
  {
    option: T;
  }
>;
export type SetTile = PayloadAction<
  typeof SET_TILE,
  {
    row: number;
    col: number;
  }
>;
export type CheckState = Action<typeof CHECK_STATE>;
export type EndGame = Action<typeof END_GAME>;

export const initGame = <T>(
  board: Board<T> = [[]],
  options: Options<T> = [],
  voidOption?: T,
): InitGame<T> => ({
  type: INIT_GAME,
  payload: {
    board,
    voidOption,
    options,
  },
});

export const setOption = <T>(option: T): SetOption<T> => ({
  type: SET_OPTION,
  payload: {
    option,
  },
});

export const setTile = (col: number = 0, row: number = 0): SetTile => ({
  type: SET_TILE,
  payload: {
    row,
    col,
  },
});

export const checkState = (): CheckState => ({
  type: CHECK_STATE,
});

export const endGame = (): EndGame => ({
  type: END_GAME,
});
