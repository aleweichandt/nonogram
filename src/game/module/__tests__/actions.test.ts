import {initGame, setOption, setTile, checkState, endGame} from '../actions';
import {Board, Options} from '../types';

const B = 'B';
const V = 'V';
type Option = typeof B | typeof V;

describe('game actions test suite', () => {
  const testBoard: Board<Option> = [[B, V], [V, B]];
  const testOptions: Options<Option> = [B, V];

  describe('init game', () => {
    it('default params', () => {
      expect(initGame()).toMatchSnapshot();
    });
    it('with params', () => {
      expect(initGame(testBoard, testOptions, V)).toMatchSnapshot();
    });
  });
  describe('set option', () => {
    it('with params', () => {
      expect(setOption(B)).toMatchSnapshot();
    });
  });
  describe('set tile', () => {
    it('default params', () => {
      expect(setTile()).toMatchSnapshot();
    });
    it('with params', () => {
      expect(setTile(1, 1)).toMatchSnapshot();
    });
  });
  it('check state', () => {
    expect(checkState()).toMatchSnapshot();
  });
  it('end game', () => {
    expect(endGame()).toMatchSnapshot();
  });
});
