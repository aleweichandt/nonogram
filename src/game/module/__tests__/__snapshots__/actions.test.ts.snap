// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game actions test suite check state 1`] = `
Object {
  "type": "game/CHECK_STATE",
}
`;

exports[`game actions test suite end game 1`] = `
Object {
  "type": "game/END_GAME",
}
`;

exports[`game actions test suite init game default params 1`] = `
Object {
  "payload": Object {
    "board": Array [
      Array [],
    ],
    "options": Array [],
    "voidOption": undefined,
  },
  "type": "game/INIT_GAME",
}
`;

exports[`game actions test suite init game with params 1`] = `
Object {
  "payload": Object {
    "board": Array [
      Array [
        "B",
        "V",
      ],
      Array [
        "V",
        "B",
      ],
    ],
    "options": Array [
      "B",
      "V",
    ],
    "voidOption": "V",
  },
  "type": "game/INIT_GAME",
}
`;

exports[`game actions test suite set option with params 1`] = `
Object {
  "payload": Object {
    "option": "B",
  },
  "type": "game/SET_OPTION",
}
`;

exports[`game actions test suite set tile default params 1`] = `
Object {
  "payload": Object {
    "col": 0,
    "row": 0,
  },
  "type": "game/SET_TILE",
}
`;

exports[`game actions test suite set tile with params 1`] = `
Object {
  "payload": Object {
    "col": 1,
    "row": 1,
  },
  "type": "game/SET_TILE",
}
`;
